### ISSUE TITLE
wsk action list for package bindings

### ISSUE THREAD

**Issue by sjfink** [openwhisk/openwhisk#1843]
Right now, wsk action list does not show the actions from package bindings of public packages.

I think this can be confusing, since wsk activation list shows the corresponding activations -- i suggest that's it would be useful if wsk action list included all actions that might appear in wsk activation list.

**Comment by rabbah** [comment]
This is entirely on the API side - generating the join for the binding requires some work in the view with complementary post processing in the controller possibly.

**Comment by style95** [comment]
I want to discuss the preferred approach for this issue.
If we keep current package binding feature, I think expected sub-tasks for this issue are as follow:

1. [Optional] Add new flag(ex: `bindings`) to determine whether bound actions should be listed in `list actions` API or not.
2. List up all bound packages in the namespace.
3. Read `binding` field from each bound packages and list up all actions with `full entity path` of reference package(binding package).
4. Change `namespace` field of each retrieved actions to the value of bound package`s one.
5. Combine all results with non-bound actions.

I think this will cause a few overhead in terms of latency.

Alternative approach is as follow:
Copy actions information when binding a package.
Bound actions may have `binding` field to be distinguishable from the others.

With this change, controller can retrieve actions transparently with `actions` view no matter whether action is bound or not.
But it will introduce more spatial overhead in CouchDB.

And it is required to add few method to handle bound entities.
For example, it is required to prevent `update API` on bound action.
And in case reference package is deleted, bound package and action should return proper error message and so on.


I am inclined to the second option, but want to hear any opinions or feedbacks

**Comment by rabbah** [comment]
If you list packages, you will find there is a binding annotation already on assets that are bindings. We could add information in the list view though to make the client side join easier.

I think the second option if I understood it changes the semantics of package bindings. From by reference to by copy. The benefits of the former is one copy vs many. If one is going to share the package say on GitHub the making copies through existing deploy mechanisms make more sense.

**Comment by style95** [comment]
Thank you for the opinion.

I got your idea.
Then how do you think this kind of approach?

We can add actions information in `binding` field.
There will be the ID of its original action.

example) bind `originPackage` in `yourNamespace` to `myPackage` in `myNamespace`

```
binding: {
  namespace: "yourNamespace",
  name: "originPackage"
 actions: [
  { 
    origin:   "yourNamespace/originPackage/action1" // action ID in CouchDB
   },
  { 
    origin:   "yourNamespace/originPackage/action2"
   }
 ]
}
namespace: myNamespace
name: myPackage
```

And in the couchDB view, we can link original action document, with those `origin` information.
With this change, we can emit the following key, value.

```
key: ["myNamespace/myPackage", ${TIMESTAMP}, "action1"]
value: {namespace: "myNamespace/myPackage", name: "action1",  ...(with all other fields in original actions)}
```
So the information of bound actions will be available in `actions` view.

**Comment by rabbah** [comment]
I don't think we should snapshot the actions in the package into the annotations (since these can change). Instead, I would favor changing the `binding` field which is computed by the view to instead of being a Boolean type to a string which is the fully qualified name of the referenced package. Then a client can list all the packages and then fetch the actions via listing actions in the package namespace.

**Comment by style95** [comment]
I agree on major part of your opinion.
But if we only add FQN of referened package, how can we differentiate bound actions with original actions in same name space?
Are you thinking of refining them at controller layer?

**Comment by style95** [comment]
So if original actions or package are changed, would it be propagated to the bound package?
If we only keep FQN of original package in bound package, we cannot link original actions in CouchDB layer. Instead we can fetch all actions in original package and refine them to change the namespace to bound package at controller layer.

**Comment by style95** [comment]
If we can get bound actions from `actions` view directly it would be great, rather than get referenced actions multiple time for each binding and combine them to respond to the client.
I will more deeply look into current view and linked document.
